File: 프로그래머스/1/42576. 완주하지 못한 선수/완주하지 못한 선수.js
// completion 배열을 순회하면서 들어온 참가자 추가하기
// participant 배열 순회하면서 참여한 참가자 제거하기
// 순회가 끝났을때 갯수가 0이 아닌 애가 완주하지 못한 선수

function solution(participant, completion) {
  const runner = new Map();
  // 1
  completion.forEach((c) => {
    runner.set(c, runner.get(c) + 1 || 1);
  });

  // 2
  for (let p of participant) {
    if (!runner.get(p)) return p;
    else runner.set(p, runner.get(p) - 1);
  }
}

console.log(solution(["leo", "kiki", "eden"], ["eden", "kiki"]));
console.log(
  solution(
    ["marina", "josipa", "nikola", "vinko", "filipa"],
    ["josipa", "filipa", "marina", "nikola"]
  )
);

console.log(
  solution(["mislav", "stanko", "mislav", "ana"], ["stanko", "ana", "mislav"])
);

File: 프로그래머스/1/42576. 완주하지 못한 선수/완주하지 못한 선수.java
import java.util.*;

class Solution {
    public String solution(String[] participant, String[] completion) {
        /*
        1. 아이디어
        참여자들을 돌면서 참여한 인원의 명수를 HashMap에 저장
        완주한자들을 돌면서 HashMap에 있으면 --;
        HashMap의 keySet을 돌면서 아직도 존재하면 그사람이 미완주자
        
        2. 시간복잡도
        O(logn)
        3. 자료구조
        HashMap
        */
        
        HashMap<String,Integer> hm = new HashMap<>();
        for(String p : participant){
            if(hm.get(p)==null) hm.put(p,0);
            hm.put(p,hm.get(p) +1);
        }
        
        for(String c : completion){
            if(hm.get(c) > 0) hm.put(c,hm.get(c) - 1);
        }

        String result="";
        for(String key : hm.keySet()){
            if(hm.get(key)>0) result= key;
        }
        return result;
    }
}